### PR TITLE
CP-2787 Docs overhaul & v3.0.0 upgrade guide

### PR DESCRIPTION
**Pull request by evanweible-wf**
## Improvement
The README was too big to be efficiently consumed. It was also slightly outdated and was missing some key examples/guides. We also needed a 3.0.0 upgrade guide.

## Changes
- Simplify the README.md to only include only immediately pertinent information for consumers.

  - Project description
  - Links to docs & guides
  - Installation/import instructions
  - Dart SDK compatibility
  - Versioning & stability
  - Credits
  - Development notes

- Extract the rest of the readme to the `/docs/` directory and split it up into multiple, consumable guides with a single table of contents.

- Add a v2.x -> v3.0.0 upgrade guide

## Testing
N/a, docs only change.

## Code Review
@trentgrover-wf 
@maxwellpeterson-wf 
@dustinlessard-wf 
@jayudey-wf 
@sebastianmalysa-wf

### CODE REVIEW

**Comment by aviary-wf** [comment]
## Raven
Number of Findings: 3
* Watched keyword ```x-xsrf-token``` in ```docs/guides/HttpTransportClient.md``` line(s) ['13'] added
* Watched keyword ```x-xsrf-token``` in ```README.md``` line(s) ['507', '586'] removed
* Watched keyword ```x-xsrf-token``` in ```docs/guides/HttpIntercepting.md``` line(s) ['107'] added

As a result of the finding(s) listed above, a security review is required for this pull request.

**Comment by rmconsole-wf** [comment]
@evanweible-wf
* It looks like you have not included any tickets in the title of this PR, but this repo requires tickets for all pull requests. Please create a ticket for this change, or click <a href='https://w-rmconsole.appspot.com/api/v1/pulls/81410604/addTicketToTitle'>here</a> to have Rosie create one for you.

**Comment by rmconsole2-wf** [comment]
Sorry, I could not create a JIRA ticket for this pull.

I sent the following data to JIRA:
`{'issuetype': {'name': u'Task'}, 'project': {'key': u'CP'}, 'description': '', 'components': [{'name': u'w_transport'}], 'summary': 'WIP: Docs overhaul & v3.0.0 upgrade guide'}`

I received the following data from JIRA:
`Hidden on Open Source Repos`

If this is an issue that you cannot resolve, please reach out to RM in the Release Management chat room.

**Comment by rmconsole2-wf** [comment]
Sorry, I could not create a JIRA ticket for this pull.

I sent the following data to JIRA:
`{'issuetype': {'name': u'Task'}, 'project': {'key': u'CP'}, 'description': '', 'components': [{'name': u'w_transport'}], 'summary': 'WIP: Docs overhaul & v3.0.0 upgrade guide'}`

I received the following data from JIRA:
`Hidden on Open Source Repos`

If this is an issue that you cannot resolve, please reach out to RM in the Release Management chat room.

File: docs/guides/HttpContentEncoding.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+### Encoding
+
+Every request instance has an `Encoding encoding` property that defaults to
+UTF8. The encoding can be set to something else (e.g. ASCII or LATIN1), but for
+most cases the default UTF8 should be sufficient.
+
+The selected encoding is significant for three reasons:
+
+- The HTTP logic in this library may use the selected encoded to encode or
```
Comment (maxwellpeterson-wf) [inline docs/guides/HttpContentEncoding.md]: ~~encoded~~ encoding

File: docs/guides/MockInstallation.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+When the transport mocks are installed, it is up to you to define how requests
+and WebSockets are handled. In other words, you play the role of the server so
+that a real one isn't necessary.
+
+To do this, you can set up **expectations** and/or **handlers**. Expectations
+are one-time only, while handlers continue to serve requests/WebSockets until
+canceled.
+
+Check out the HTTP and WebSocket guides for expectations and handlers.
```
Comment (maxwellpeterson-wf) [inline docs/guides/MockInstallation.md]: Link?

File: docs/guides/MockInstallation.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+import 'package:w_transport/mock.dart';
+
+void main() {
+  setUp(() {
+    MockTransports.install();
+  });
+  
+  tearDown(() {
+    MockTransports.verifyNoOutstandingExceptions();
+    MockTransports.uninstall();
```
Comment (maxwellpeterson-wf) [inline docs/guides/MockInstallation.md]: Does this need to be awaited?

File: docs/guides/MockWebSocketHandlers.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    Uri.parse('/ws'),
+    (Uri uri, {Iterable<String> protocols, Map<String, dynamic> headers}) async {
+      // Perform any setup necessary.
+      // Can even dynamically return a mock WebSocket server based on the given
+      // URI, protocols, and headers.
+      return mockWebSocketServer;
+    });
+```
+
+### Registering a handler to match WebSockets a URI pattern
```
Comment (maxwellpeterson-wf) [inline docs/guides/MockWebSocketHandlers.md]: Not 100% sure how this should read but I think:

```diff
- Registering a handler to match WebSockets a URI pattern
+ Registering a handler to match a WebSockets URI pattern
```

Comment (evanweible-wf) [inline docs/guides/MockWebSocketHandlers.md]: Fixed

File: docs/guides/MockInstallation.md (hunk 3 above)
Comment (evanweible-wf) [inline docs/guides/MockInstallation.md]: Good catch, fixed

File: docs/guides/MockInstallation.md (hunk 2 above)
Comment (evanweible-wf) [inline docs/guides/MockInstallation.md]: Added

File: docs/guides/HttpContentEncoding.md (hunk 1 above)
Comment (evanweible-wf) [inline docs/guides/HttpContentEncoding.md]: fixed

**Comment by maxwellpeterson-wf** [comment]
+1

**Comment by dustinlessard-wf** [comment]
+1

**Comment by sebastianmalysa-wf** [comment]
+1

**Comment by maxwellpeterson-wf** [comment]
QA +1
- [x] Testing instruction
- [x] Dev +1's
- [x] Dev/QA +10 with detail of what was tested
  - Docs only change
- [ ] Unit tests created/updated 
  - Docs-only change
- [x] All unit tests pass

Merging into master.

**Comment by maxwellpeterson-wf** [comment]
+1 security

**Comment by evanweible-wf** [comment]
@rmconsole-wf run_merge_script

**Pull request merged**